File: Design War Measurement System/Code/Java/src/fayelab/ddd/measure/vector/MeasureSystem.java
package fayelab.ddd.measure.vector;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.IntStream;

public class MeasureSystem
{
    private List<Integer> stepFactors;
    private List<Integer> baseFactors;

    public MeasureSystem(List<Integer> stepFactors)
    {
        this.stepFactors = stepFactors;
        this.baseFactors = makeBaseFactors(stepFactors);
    }

    public int base(List<Integer> quantityVec)
    {
        return VecOp.dotProduct(quantityVec, baseFactors);
    }

    public List<Integer> normalize(List<Integer> quantityVec)
    {
        return VecOp.div(quantityVec, stepFactors);
    }

    public boolean equal(List<Integer> quantityVec1, List<Integer> quantityVec2)
    {
        return VecOp.equal(normalize(quantityVec1), normalize(quantityVec2));
    }

    public List<Integer> add(List<Integer> quantityVec1, List<Integer> quantityVec2)
    {
        return normalize(VecOp.add(quantityVec1, quantityVec2));
    }

    public List<Integer> scale(int c, List<Integer> quantityVec)
    {
        return normalize(VecOp.scale(c, quantityVec));
    }

    private List<Integer> makeBaseFactors(List<Integer> stepFactors)
    {
        List<Integer> result = new ArrayList<>();
        IntStream.range(1, stepFactors.size())
                 .collect(() -> result,
                          (acc, idx) -> {
                              List<Integer> tail = stepFactors.subList(idx, stepFactors.size());
                              acc.add(tail.stream().reduce(1, (x, y) -> x * y));
                          },
                          (acc1, acc2) -> {
                              acc1.addAll(acc2);
                          });
        result.add(1);
        return result;
    }
    
    List<Integer> getStepFactors()
    {
        return stepFactors;
    }

    List<Integer> getBaseFactors()
    {
        return baseFactors;
    }
}

File: DDD DSL/DDD DSL Design War Measurement System/Code/Python/vector/measuresystem.py
import vecop

def basefactors(stepfactors):
    result = []
    for (i, stepFactor) in enumerate(stepfactors[1:]):
        result.append(reduce(lambda x, y: x * y, stepfactors[i + 1:]))
    result.append(1)
    return result

def basefactors2(stepfactors):
    return reduce(_acc_basefactors2, stepfactors[::-1][:-1], [1])[::-1]

def _acc_basefactors2(acc, stepfactor):
    acc.append(acc[-1] * stepfactor)
    return acc

def base(quantityvec, basefactors):
    return vecop.dotproduct(quantityvec, basefactors)

def normalize(quantityvec, stepfactors):
    return vecop.div(quantityvec, stepfactors)

def equal(quantityvec1, quantityvec2, stepfactors):
    return vecop.equal(normalize(quantityvec1, stepfactors),
                       normalize(quantityvec2, stepfactors))

def add(quantityvec1, quantityvec2, stepfactors):
    return normalize(vecop.add(quantityvec1, quantityvec2), stepfactors)

def scale(c, quantityvec, stepfactors):
    return normalize(vecop.scale(c, quantityvec), stepfactors)